engine.lisp

```diff
@@ -87,6 +87,90 @@ for the selector to the matcher."))
                      (incf end)
                   finally (test) (return NIL))))))))
 
+;; https://drafts.csswg.org/cssom/#serialize-an-identifier
+(defun css-escape (string)
+  "Escape all the invalid CSS characters to their safe counterparts."
+  (declare (optimize speed)
+           (type string string))
+  (let ((buffer (make-array 0 :adjustable t :fill-pointer 0)))
+    (labels ((escapable (char)
+               (let ((code (char-code char)))
+                 (or (<= code #x1f)
+                     (= code #x7f)
+                     (not (or (>= code #x80)
+                              (char= char #\-)
+                              (char= char #\_)
+                              (digit-char-p char)
+                              (char<= #\A char #\Z)
+                              (char<= #\a char #\z))))))
+             (extend-with-string (string)
+               (loop for char across string
+                     do (vector-push-extend char buffer)))
+             (extend-with-escaped (char)
+               (extend-with-string (format nil "\\~X " (char-code char))))
```

Comment:
Using a stream would also avoid consing these extra strings here for no good reason.